### DISCUSSION TITLE
Transforming HL7 v2.x to JSON with Mirth (open-source project)

### DISCUSSION THREAD

**Discussion by ropese34** [nextgenhealthcare/connect#5443] [Q&A] [answered]
Hello,

We are using the open-source project version of Mirth. Could someone outline what is needed to transform incoming V2.x HL7 to JSON or MQTT messages, transform from HL7 to JSON messages and send the mapped messages to something like a destination SQS queue?
Would this require a custom library? 

Thank you

**Comment by pacmano1** [comment, marked as answer]
Transforming HL7 to JSON is standard and common for Mirth.   Please use Google to find examples. There are video walk throughs on YouTube.  No additional libraries are needed.

AWS SQS supports JMS and Mirth supports JMS, so that should work as well.     However using the java library would be another option.   So Yes and No to your last question.

Please also visit the mirth forums for many years of dicussions with helpful information.

**Reply by ropese34** [reply]
Thank you. I think I am close to a solution. Thanks to the information readily available online. My boss wants to have something run nightly to essentially transform the HL7 messages received during that day and write a single .json file to our AWS s3. I don't have the permissions yet to the s3 and the specific location/bucket so for now, I just want to see the output in the appropriate Mirth channel log. I created a new channel with Source Connector Inbound and Outbound data type set to HL7 v2.x. Destination1  Inbound is HL7 v2.x datatype and Outbound is JSON datatype. In the channel properties, I set source temporarily to Channel Reader (I want to transform the raw hl7 that comes in on the "Source" destination from the source system, into JSON). I set Destination connector type to Channel Writer. In the Destination Transformer, I created a HL7 to JSON javascript writer step and added appropriate code. I pasted a sample SIU in the Inbound message template tree and Mirth converted it to JSON structure for the Outbound message template tree. In the channel properties of existing device that receives the source hl7, I added the HL7 to JSON destination I created and move it to the top of the destinations list. I then reprocessed a message that came in through the Source or default. I don't see anything received or processed by the new channel I created.

**Reply by pacmano1** [reply]
Post code.
Post screenshots.  
Post a full export of your channel.

Some of your terminology does not make sense.   Like "existing device", so pictures and code would be more helpful.

**Reply by ropese34** [reply]
Coming up. I didn't even notice I used terminology from the previous engine I worked with. Wow, sorry about that.

**Reply by ropese34** [reply]
![Channelsjpg](https://user-images.githubusercontent.com/108378567/194146359-e10e95a4-1cf3-436e-bd3f-1b202dd02cf7.jpg)
![ChannelProperties](https://user-images.githubusercontent.com/108378567/194146412-aa40566c-12e7-484d-8af3-f6dd6e7fb36a.jpg)
![Channel_Source](https://user-images.githubusercontent.com/108378567/194146437-f9b1a551-7bd3-4d3b-9b80-1c892fda3500.jpg)
![Channel_Destinations](https://user-images.githubu
![ChannelDestination1Transformer](https://user-images.githubusercontent.com/108378567/194146551-75dfa646-418b-4378-ae9d-5bf00b46ff3a.jpg)
sercontent.com/108378567/194146498-19f80d34-d779-4a9e-adc5-8adb8fe1e42c.jpg)

**Reply by pacmano1** [reply]
Well done!

So a simple version:

![image](https://user-images.githubusercontent.com/44065187/194147698-5fcb660b-992c-40fb-9d65-de1068867e64.png)

![image](https://user-images.githubusercontent.com/44065187/194147974-45ae16f1-cb1f-4407-9f02-1944a8a3ca8a.png)

**Reply by ropese34** [reply]
![Error](https://user-images.githubusercontent.com/108378567/194149401-fa2016f2-c09e-4622-97ec-5425d76b2f4f.jpg)

**Reply by ropese34** [reply]
> Well done!

Really ?!? Means a lot coming from you. I will review in detail right now. Thank you!!

**Reply by pacmano1** [reply]
See attached.  This is about as simple as can be for an hl7 to json channel.  Rename the attached to .xml then import.

Try to send the hl7 message I have in the inbound transformer message template into this channel after you import the channel.  Ignore the premium plugin warnings you may see on import.

[hl7 in JSON out.txt](https://github.com/nextgenhealthcare/connect/files/9718594/hl7.in.JSON.out.txt)

**Reply by ropese34** [reply]
Thanks. Did you mean I made it more complicated that it should be or something obvious that I missed? I will go through 
this shortly...

**Reply by ropese34** [reply]
I'm on an older version of Mirth than you.

**Reply by pacmano1** [reply]
Fire up a docker container or VM with my version or later and import it.

Or post your full channel export.

**Reply by ropese34** [reply]
What am I missing from my solution? I will have to work with our team to upgrade or Mirth instances.

**Reply by ropese34** [reply]
> Fire up a docker container or VM with my version or later and import it.
> 
> Or post your full channel export.

include library in the export, right? coming right up...

**Reply by ropese34** [reply]
[HL7v2x to JSON_ropese34.txt](https://github.com/nextgenhealthcare/connect/files/9718727/HL7v2x.to.JSON_ropese34.txt)

**Reply by pacmano1** [reply]
nah, library not needed.  You should not have to call any code template libraries for this.

**Reply by ropese34** [reply]
let me know and i can export again. i did include the library it said was linked. thanks

**Reply by pacmano1** [reply]
OK:

1) reset the data types to the defauls, on this screen:

![image](https://user-images.githubusercontent.com/44065187/194156142-ea6fca80-9e25-4778-ac17-b67ad2a21e35.png)

2) I expect your assigments are not what you want, you need to explicity parse the HL7 message and assign each element to each JSON element otherwise your code just inserts  the XML representation of the segment into the JSON blob:

![image](https://user-images.githubusercontent.com/44065187/194156533-6cf5c826-8076-4b56-979c-c66507428475.png)

My guess is that is not what your task is.

**Reply by pacmano1** [reply]
That means you will be iterating over hl7 constructs AND building JSON elements and arrays of elements to represent that data.

Whoever asked you to do this should have said "I want the JSON to look like this".

**Reply by ropese34** [reply]
I should be able to map the full corresponding HL7 segment to each JSON element, right?

**Reply by ropese34** [reply]
My data type selections, match yours:
![ChannelProperties](https://user-images.githubusercontent.com/108378567/194157731-4eb3df22-95dd-4c8d-8332-813fb238101b.jpg)

**Reply by ropese34** [reply]
One of your screenshot shows two destinations.

**Reply by pacmano1** [reply]
No, it shows the message sent twice.

**Reply by ropese34** [reply]
Ahh, yes, right. Twice, same destination. Ok so my data type selections were correct? It's my javascript thats wrong. I'm essentially not mapping the hl7 to json elements correctly at all it looks like, right?

**Reply by pacmano1** [reply]
You data type selections don't match mine (note "strip namespace"), that was a change in the defaults a few releases ago.

And yes, your mapping is not correct - it will product output, but I very much doubt that is your desired output.  

Aagin = whoever asked you to do this should have said "I want the JSON to look like this".

**Comment by ropese34** [comment]
I've had to table this task but am revisiting now. I made some changes to for now output/write source messages to channel log but am not seeing any messages on the destination

![OutputChannel_Destination](https://user-images.githubusercontent.com/108378567/198071631-84d26c33-acc8-4803-a4c6-4a83261110e2.jpg)
![OutputChannel_Source](https://user-images.githubusercontent.com/108378567/198071633-18f09d40-9e07-4b82-acc9-fd0b95525d9b.jpg)
![SIU_ORM_channel_properties](https://user-images.githubusercontent.com/108378567/198071638-c04ab0f5-8dd5-4ac5-93e4-0ebe8a622e75.jpg)
![Dash](https://user-images.githubusercontent.com/108378567/198071639-12b11bdd-d14e-4f6e-ba13-bc2e28f83a81.jpg)

**Comment by tom08zehn** [comment]
Why don't you simply do: `tmp = XmlUtil.toJson(msg);`

Explanation:

## Convert HL7 to JSON using Mirth

Mirth is parsing any inbound message and represents it internally in E4X (XML) format (except RAW or binary message types). The following code transforms E4X/XML to JSON.

### Prerequisites

1.	Edit channel -> *Summary* tab -> *Data Types*: click button *Set Data Types*
	*	Source Connector
		*	Inbound: HL7 v2.x
		*	Outbound: JSON
2.	Recommended: *Source* tab
	*	Connector Type: TCP Listener
	*	Source Settings
		*	Source Queue: ON (Responds before processing), ensures order of incoming messages
		*	Response: Auto-generate (Before processing), immediately sends an HL7/ACK back to sender and doesn't return any errors (sender believes that everything is fine)
	*	Configure other settings such as *Local Port*, *Encoding*, etc.
3.	Edit channel -> *Source* tab -> from *Channel Tasks* (left pane), select *Edit Transformer*
	*	From *Transformer Tasks* (left pane), select *Add New Step* -> change the *Type* to *JavaScript*
	*	Select *Message Templates* tab (right pane)
		*	Unter *Outbound Message Template* (*Data Type* shows *JSON*), enter `{}` (empty object), otherwise the `tmp` variable is unset in the transformer script

### Transformer script code

```js
tmp = XmlUtil.toJson(msg);
```

### Check

Send the following example ORM message:

```hl7
MSH|^~&|HIS|MedCenter|LIS|MedCenter|20060307110114||ORM^O01|MSGID20060307110114|P|2.3
PID|||12001||Jones^John^^^Mr.||19670824|M|||123 West St.^^Denver^CO^80020^USA|||||||
PV1||O|OP^PAREG^||||2342^Jones^Bob|||OP|||||||||2|||||||||||||||||||||||||20060307110111|
ORC|NW|20060307110114
OBR|1|20060307110114||003038^Urinalysis^L|||20060307110114
```

Encoded (and transformed) message is:

```json
{
  "HL7Message" : {
    "MSH" : {
      "MSH.1" : "|",
      "MSH.2" : "^~\\&",
      "MSH.3" : {
        "MSH.3.1" : "HIS"
      },
      "MSH.4" : {
        "MSH.4.1" : "MedCenter"
      },
      "MSH.5" : {
        "MSH.5.1" : "LIS"
      },
      "MSH.6" : {
        "MSH.6.1" : "MedCenter"
      },
      "MSH.7" : {
        "MSH.7.1" : 20060307110114
      },
      "MSH.8" : null,
      "MSH.9" : {
        "MSH.9.1" : "ORM",
        "MSH.9.2" : "O01"
      },
      "MSH.10" : {
        "MSH.10.1" : "MSGID20060307110114"
      },
      "MSH.11" : {
        "MSH.11.1" : "P"
      },
      "MSH.12" : {
        "MSH.12.1" : 2.3
      }
    },
    "PID" : {
      "PID.1" : null,
      "PID.2" : null,
      "PID.3" : {
        "PID.3.1" : 12001
      },
      "PID.4" : null,
      "PID.5" : {
        "PID.5.1" : "Jones",
        "PID.5.2" : "John",
        "PID.5.3" : null,
        "PID.5.4" : null,
        "PID.5.5" : "Mr."
      },
      "PID.6" : null,
      "PID.7" : {
        "PID.7.1" : 19670824
      },
      "PID.8" : {
        "PID.8.1" : "M"
      },
      "PID.9" : null,
      "PID.10" : null,
      "PID.11" : {
        "PID.11.1" : "123 West St.",
        "PID.11.2" : null,
        "PID.11.3" : "Denver",
        "PID.11.4" : "CO",
        "PID.11.5" : 80020,
        "PID.11.6" : "USA"
      },
      "PID.12" : null,
      "PID.13" : null,
      "PID.14" : null,
      "PID.15" : null,
      "PID.16" : null,
      "PID.17" : null,
      "PID.18" : null
    },
    "PV1" : {
      "PV1.1" : null,
      "PV1.2" : {
        "PV1.2.1" : "O"
      },
      "PV1.3" : {
        "PV1.3.1" : "OP",
        "PV1.3.2" : "PAREG",
        "PV1.3.3" : null
      },
      "PV1.4" : null,
      "PV1.5" : null,
      "PV1.6" : null,
      "PV1.7" : {
        "PV1.7.1" : 2342,
        "PV1.7.2" : "Jones",
        "PV1.7.3" : "Bob"
      },
      "PV1.8" : null,
      "PV1.9" : null,
      "PV1.10" : {
        "PV1.10.1" : "OP"
      },
      "PV1.11" : null,
      "PV1.12" : null,
      "PV1.13" : null,
      "PV1.14" : null,
      "PV1.15" : null,
      "PV1.16" : null,
      "PV1.17" : null,
      "PV1.18" : null,
      "PV1.19" : {
        "PV1.19.1" : 2
      },
      "PV1.20" : null,
      "PV1.21" : null,
      "PV1.22" : null,
      "PV1.23" : null,
      "PV1.24" : null,
      "PV1.25" : null,
      "PV1.26" : null,
      "PV1.27" : null,
      "PV1.28" : null,
      "PV1.29" : null,
      "PV1.30" : null,
      "PV1.31" : null,
      "PV1.32" : null,
      "PV1.33" : null,
      "PV1.34" : null,
      "PV1.35" : null,
      "PV1.36" : null,
      "PV1.37" : null,
      "PV1.38" : null,
      "PV1.39" : null,
      "PV1.40" : null,
      "PV1.41" : null,
      "PV1.42" : null,
      "PV1.43" : null,
      "PV1.44" : {
        "PV1.44.1" : 20060307110111
      },
      "PV1.45" : null
    },
    "ORC" : {
      "ORC.1" : {
        "ORC.1.1" : "NW"
      },
      "ORC.2" : {
        "ORC.2.1" : 20060307110114
      }
    },
    "OBR" : {
      "OBR.1" : {
        "OBR.1.1" : 1
      },
      "OBR.2" : {
        "OBR.2.1" : 20060307110114
      },
      "OBR.3" : null,
      "OBR.4" : {
        "OBR.4.1" : "003038",
        "OBR.4.2" : "Urinalysis",
        "OBR.4.3" : "L"
      },
      "OBR.5" : null,
      "OBR.6" : null,
      "OBR.7" : {
        "OBR.7.1" : 20060307110114
      }
    }
  }
}
```

### Logging

You can log it in-line as follows: `logger.debug(tmp);`

Or prettyfied: `logger.debug(JsonUtil.prettyPrint(tmp));`

**Reply by pacmano1** [reply]
Well, sure but most JSON endpoints don't take data labeled like that.

**Reply by tom08zehn** [reply]
[ropese34](https://github.com/ropese34) just asked for JSON conversion... the result is valid JSON.

If you mean because the property names contain dots: You could simply replace all dots by hyphens with a single line of code using `JSON.stringify()` and the [replacer argument](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON/stringify?retiredLocale=de#the_replacer_parameter).

^^ I recommend this approach if you simply want to convert *the whole* HL7 message to JSON, e.g. in order to forward it to an HTTP(S) endpoint.

If you need to transform HL7 to a specific JSON template, then you could simply do it as shown above, like:

```js
tmp.patientFirstName = msg['PID']['PID.5']['PID.5.2'];
tmp.patientLastName = msg['PID']['PID.5']['PID.5.1'];
logger.debug(JSON.stringify(tmp));
```

Log output:

```
{
  "patientFirstName": "John",
  "patientLastName": "Doe"
}
```

**Reply by pacmano1** [reply]
`msg['PID']['PID.5']['PID.5.2'].toString()` I think you mean.
